agent/consul/leader_connect_ca.go

```diff
@@ -0,0 +1,1082 @@
+	newActiveRoot.IntermediateCerts = append(newActiveRoot.IntermediateCerts, intermediatePEM)
+	newActiveRoot.SigningKeyID = connect.EncodeSigningKeyID(intermediateCert.SubjectKeyId)
+
+	c.logger.Info("received new intermediate certificate from primary datacenter")
+	return nil
+}
+
+// intermediateCertRenewalWatch periodically attempts to renew the intermediate cert.
+func (c *CAManager) intermediateCertRenewalWatch(ctx context.Context) error {
+	isPrimary := c.srv.config.Datacenter == c.srv.config.PrimaryDatacenter
+
+	for {
+		select {
+		case <-ctx.Done():
+			return nil
+		case <-time.After(structs.IntermediateCertRenewInterval):
+			retryLoopBackoffAbortOnSuccess(ctx, func() error {
+				return c.RenewIntermediate(isPrimary)
+			}, func(err error) {
+				c.logger.Error("error renewing intermediate certs",
+					"routine", intermediateCertRenewWatchRoutineName,
+					"error", err,
+				)
+			})
+		}
+	}
+}
+
+// RenewIntermediate checks the intermediate cert for
+// expiration. If more than half the time a cert is valid has passed,
+// it will try to renew it.
+func (c *CAManager) RenewIntermediate(isPrimary bool) error {
+	// Grab the 'lock' right away so the provider/config can't be changed out while we check
+	// the intermediate.
+	if err := c.setState(CAStateRenewIntermediate, true); err != nil {
+		return err
+	}
+	defer c.setState(CAStateReady, false)
+
+	provider, _ := c.getCAProvider()
+	if provider == nil {
+		// this happens when leadership is being revoked and this go routine will be stopped
+		return nil
+	}
+	// If this isn't the primary, make sure the CA has been initialized.
+	if !isPrimary && !c.configuredSecondaryCA() {
+		return fmt.Errorf("secondary CA is not yet configured.")
+	}
+
+	state := c.srv.fsm.State()
+	_, activeRoot, err := state.CARootActive(nil)
+	if err != nil {
+		return err
+	}
+
+	// If this is the primary, check if this is a provider that uses an intermediate cert. If
+	// it isn't, we don't need to check for a renewal.
+	if isPrimary {
+		_, config, err := state.CAConfig(nil)
+		if err != nil {
+			return err
+		}
+
+		if _, ok := ca.PrimaryIntermediateProviders[config.Provider]; !ok {
+			return nil
+		}
+	}
+
+	activeIntermediate, err := provider.ActiveIntermediate()
+	if err != nil {
+		return err
+	}
+
+	if activeIntermediate == "" {
+		return fmt.Errorf("datacenter doesn't have an active intermediate.")
+	}
+
+	intermediateCert, err := connect.ParseCert(activeIntermediate)
+	if err != nil {
+		return fmt.Errorf("error parsing active intermediate cert: %v", err)
+	}
+
+	if lessThanHalfTimePassed(time.Now(), intermediateCert.NotBefore.Add(ca.CertificateTimeDriftBuffer),
+		intermediateCert.NotAfter) {
+		return nil
+	}
+
+	// Enough time has passed, go ahead with getting a new intermediate.
+	renewalFunc := c.getIntermediateCAPrimary
+	if !isPrimary {
+		renewalFunc = c.getIntermediateCASigned
+	}
+	if err := renewalFunc(provider, activeRoot); err != nil {
+		return err
+	}
```

Comment:
In a secondary this is going to do an RPC over the wan which may take a long time. From the perspective of the secondary, this operation itself doesn't do any local writes so you could do the following to mitigate:

1. pass the `ctx` in to this function
2. use another goroutine to invoke `renewalFunc`, shipping back the err/no-err via a `chan error`
3. here do a select on the `ctx.Done()` and (2)'s error channel so the leader routine here can cancel and get out of this locked section quickly